Userland/Libraries/LibCompress/Brotli.cpp

```diff
@@ -0,0 +1,920 @@
+    while (i < alphabet_size) {
+        auto symbol = TRY(temp_code.read_symbol(m_input_stream));
+
+        if (symbol < 16) {
+            result_symbols.append(i);
+            result_lengths.append(symbol);
+            result_lengths_count[symbol]++;
+
+            if (symbol != 0) {
+                previous_non_zero_code_length = symbol;
+                sum += (32768 >> symbol);
+                if (sum == 32768)
+                    break;
+                else if (sum > 32768)
+                    return Error::from_string_literal("invalid prefix code");
+            }
+
+            last_repeat = 0;
+            i++;
+        } else if (symbol == 16) {
+            size_t repeat_count = 0;
+            if (last_symbol == 16 && last_repeat != 0) {
+                repeat_count = (4 * (last_repeat - 2));
+            } else {
+                last_repeat = 0;
+            }
+            repeat_count += 3 + TRY(m_input_stream.read_bits(2));
+
+            for (size_t rep = 0; rep < (repeat_count - last_repeat); rep++) {
+                result_symbols.append(i);
+                result_lengths.append(previous_non_zero_code_length);
+                result_lengths_count[previous_non_zero_code_length]++;
+
+                if (previous_non_zero_code_length != 0) {
+                    sum += (32768 >> previous_non_zero_code_length);
+                    if (sum == 32768)
+                        break;
+                    else if (sum > 32768)
+                        return Error::from_string_literal("invalid prefix code");
+                }
+
+                i++;
+                if (i >= alphabet_size)
+                    break;
+            }
+            if (sum == 32768)
+                break;
+            VERIFY(sum < 32768);
+
+            last_repeat = repeat_count;
+        } else if (symbol == 17) {
+            size_t repeat_count = 0;
+            if (last_symbol == 17 && last_repeat != 0) {
+                repeat_count = (8 * (last_repeat - 2));
+            } else {
+                last_repeat = 0;
+            }
+            repeat_count += 3 + TRY(m_input_stream.read_bits(3));
+
+            i += (repeat_count - last_repeat);
+            last_repeat = repeat_count;
+        }
+
+        last_symbol = symbol;
+    }
+    result_lengths_count[0] = 0;
+
+    size_t code_value = 0;
+    for (size_t bits = 1; bits < 16; bits++) {
+        code_value = (code_value + result_lengths_count[bits - 1]) << 1;
+        size_t current_code_value = code_value;
+
+        for (size_t n = 0; n < result_symbols.size(); n++) {
+            size_t len = result_lengths[n];
+            if (len == bits) {
+                code.m_symbol_codes.append((1 << bits) | current_code_value);
+                code.m_symbol_values.append(result_symbols[n]);
+                current_code_value++;
+            }
+        }
+    }
+
+    return {};
+}
+
+static void inverse_move_to_front_transform(Span<u8> v)
+{
+    // RFC 7932 section 7.3
```

Comment:
These could be elsewhere as well, probably? (I know FLAC also needs spec comments, I'll do that at some point...)

Reply:
Yeah, some more references to the RFC might be nice.

This specific function has a reference since it was defined in the specification.

Reply:
If it's an RFC, it's basically public domain.